src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,466 @@
+                if (o == null || getClass() != o.getClass()) return false;
+                ProjPreds projPreds = (ProjPreds) o;
+                return num == projPreds.num && pred.equals(projPreds.pred) && mode.equals(projPreds.mode);
+            }
+
+            @Override
+            public int hashCode() {
+                return Objects.hash(pred, num, mode);
+            }
+        }
+
+        @RequiredArgsConstructor
+        private class NodePreds {
+            private final Node[] preds;
+            private final binding_irnode.ir_opcode opcode;
+            private final Mode mode;
+
+            @Override
+            public boolean equals(Object o) {
+                if (this == o) return true;
+                if (o == null || getClass() != o.getClass()) return false;
+                NodePreds nodePreds = (NodePreds) o;
+                return Arrays.equals(preds, nodePreds.preds) && opcode == nodePreds.opcode && mode.equals(nodePreds.mode);
+            }
+
+            @Override
+            public int hashCode() {
+                int result = Objects.hash(opcode, mode);
+                result = 31 * result + Arrays.hashCode(preds);
+                return result;
+            }
+        }
+
+        @Override
+        public void defaultVisit(Node node) {
+            // change nothing
+        }
+
+
+        private void visitPreds(Node node, Node[] preds, binding_irnode.ir_opcode opcode) {
+            NodePreds np = new NodePreds(preds, opcode, node.getMode());
+            if (nodeCache.containsKey(np)) {
+                Node n = nodeCache.get(np);
+                if (dominates(n.getBlock().ptr, node.getBlock().ptr)) {
+                    nodeValues.put(node, n);
+                } else {
+                    // replace and update cache
+                    nodeValues.put(n, node);
+                    nodeCache.put(np, node);
+                }
+            } else {
+                nodeCache.put(np, node);
+            }
+        }
+
+
+        @Override
+        public void visit(Add node) {
+            visitPreds(node, Util.iterableToArray(node.getPreds()), binding_irnode.ir_opcode.iro_Add);
+        }
+
+        @Override
+        public void visit(And node) {
+            visitPreds(node, Util.iterableToArray(node.getPreds()), binding_irnode.ir_opcode.iro_And);
+        }
+
+        @Override
+        public void visit(Cmp node) {
+            CmpPreds cmpPreds = new CmpPreds(node.getLeft(), node.getRight(), node.getRelation(), node.getMode());
+            if (cmpCache.containsKey(cmpPreds)) {
+                Node n = cmpCache.get(cmpPreds);
+                if (dominates(n.getBlock().ptr, node.getBlock().ptr)) {
+                    nodeValues.put(node, n);
+                } else {
+                    nodeValues.put(n, node);
+                    cmpCache.put(cmpPreds, node);
+                }
+            } else {
+                cmpCache.put(cmpPreds, node);
+            }
+        }
```

Comment:
You're not actually doing anything with this information. Also, I don't see any value in eliminating `Cmp`s.

Reply:
In addition, `CmpPreds` does not define an `equals` method, so `cmpCache.containsKey(cmdPreds)` will always be `false`.

Reply:
The nodeValues is updated which exchanges `Cmp` nodes, I have added equals and hashCode here though.